### DISCUSSION TITLE
pytest vs. python -m pytest

### DISCUSSION THREAD

**Discussion by RolfBippus** [pytest-dev/pytest#9753] [Q&A]
The documentation states

> Running pytest with pytest [...] instead of python -m pytest [...] yields nearly equivalent behaviour, except that the latter will add the current directory to sys.path, which is standard python behavior.

This does not seem to be the whole story.
Unfortunately, calling plain `pytest` seems to prepend the path where pytest is located to sys.path.

Printing sys.path in pytest_configure(), I get in a conda environment:

```
/home/rolf/devel > export PYTHONPATH=/home/rolf/devel/scripts
/home/rolf/devel > pytest
sys.path = [ '/home/rolf/miniforge3/envs/devel/bin', '/home/rolf/devel/scripts',  ...]
```

```
/home/rolf/devel > python -m pytest
sys.path = [ '/home/rolf/devel', '/home/rolf/devel/scripts', ...]
```

The latter is to be expected, the first not?! 

The unfortunate thing is, I now want to test a script in my development code and import it (e.g. to call a main() method) instead of running it via subprocess. The script however is also installed in my conda environment, under bin.

Calling `python -m pytest` thus imports the development code

Whereas calling plain `pytest` imports the code from the environment.

**Comment by RonnyPfannschmidt** [comment]
that the bin folder ends in sys.path looks like a bug in conda
the general suggestion is to do editable installs if you want to test editable code
the thing with `python -m` is unfixable fomr the pytest side

**Comment by RolfBippus** [comment]
> that the bin folder ends in sys.path looks like a bug in conda

the pytest script in conda calls pytest.console_main.

```
python -c "import pytest; pytest.console_main()"
sys.path = [ '', '/home/rolf/devel/scripts',  ...]
```
prepends the empty string to sys.path. So the problem seems to be calling `pytest.console_main()` ?

**Reply by RonnyPfannschmidt** [reply]
Looks like a very wrong python start script, this is a conda issue, not a pytest issue

**Reply by RolfBippus** [reply]
Thanks, it seems like the problem is related to python (not conda) and being debated for a long time now.
So indeed, you are right, the problem is not related to pytest and just by chance interferred with my setup.

It is however a problem that may pop up especially with pytest which heavily relies on the python import mechanism in order to actually work as expected.

I searched a bit and stumbled across this:

https://bugs.python.org/issue42483  and followed up on some other feeds.

Indeed, using a shell script with a python shebang actually adds the "current working dir" which in this case is the dir where the script resides (e.g. /usr/bin) to sys.path. It is the empty path if the script is in the current working dir (using e.g. python -c as I did above).

(see also https://discuss.python.org/t/python-flag-envvar-to-not-put-current-directory-on-sys-path-but-dont-ignore-pythonpath/4235 where this is discussed along with pytest).

**Reply by nicoddemus** [reply]
> Looks like a very wrong python start script, this is a conda issue, not a pytest issue

conda actually relies on `pip` to create the entry points:

https://github.com/conda-forge/pytest-feedstock/blob/91cf68174928c757348f9f5884527dc92fd7a876/recipe/meta.yaml#L15-L17

```yaml
  entry_points:
    - pytest = pytest:console_main
    - py.test = pytest:console_main
```

which is the same configuration we have in pytest:

https://github.com/pytest-dev/pytest/blob/d5ce9f5a166a5e70eeacf4b49e9fc9d93a860352/setup.cfg#L61-L64
